lib/src/main/java/org/veupathdb/lib/jaxrs/raml/multipart/MultipartMessageBodyReader.kt

```diff
@@ -0,0 +1,238 @@
+package org.veupathdb.lib.jaxrs.raml.multipart
+
+import com.fasterxml.jackson.databind.ObjectMapper
+import jakarta.ws.rs.BadRequestException
+import jakarta.ws.rs.InternalServerErrorException
+import jakarta.ws.rs.core.MediaType
+import jakarta.ws.rs.core.MultivaluedMap
+import jakarta.ws.rs.ext.MessageBodyReader
+import org.apache.commons.fileupload.MultipartStream
+import org.veupathdb.lib.jaxrs.raml.multipart.utils.*
+import java.io.File
+import java.io.InputStream
+import java.lang.reflect.Type
+
+private const val DefaultFileName = "upload"
+
+private const val BufferSize = 8192
+
+class MultipartMessageBodyReader : MessageBodyReader<Any> {
+
+  /**
+   * Tests whether this `MessageBodyReader` should kick in for the given
+   * request.
+   *
+   * @return `true` if the input request is of type `multipart/form-data`,
+   * otherwise `false`.
+   */
+  override fun isReadable(
+    type: Class<*>,
+    genericType: Type,
+    annotations: Array<out Annotation>,
+    mediaType: MediaType,
+  ) = mediaType.isCompatible(MediaType.MULTIPART_FORM_DATA_TYPE)
+
+  override fun readFrom(
+    type: Class<Any>,
+    genericType: Type,
+    annotations: Array<out Annotation>,
+    mediaType: MediaType,
+    httpHeaders: MultivaluedMap<String, String>,
+    entityStream: InputStream,
+  ): Any {
+    // Open a stream over the contents of the multipart/form-data body.
+    val stream = MultipartStream(entityStream, mediaType.requireBoundaryBytes(), BufferSize, null)
+
+    // Get hold of the temp directory assigned to this multipart request.
+    val tmpDir = httpHeaders.getTempDirectory()
```

Comment:
I think it's up to stylistic choice?  Or rather, in practice I see it every which way.  I've been personally deciding based on context and purpose.

In this case, what I _wanted_ was a getter on the header map that gives me the specific target value.  If i had been performing some complex operation on the map or mutating it in some way, I would have taken the map as a standard parameter rather than a receiver.

Reply:
From the official coding conventions:

> Use extension functions liberally. Every time you have a function that works primarily on an object, consider making it an extension function accepting that object as a receiver. To minimize API pollution, restrict the visibility of extension functions as much as it makes sense. As necessary, use local extension functions, member extension functions, or top-level extension functions with private visibility.

Reply:
Interesting.  Thanks.